### ISSUE TITLE
Displaying markdown code blocks from IJulia kernel routinely drops trailing parenthesis

### ISSUE THREAD

**Issue by goerz** [jupyterlab/jupyterlab#16112]
I Julia notebooks (using the [IJulia](https://github.com/JuliaLang/IJulia.jl) kernel), I am finding that displaying markdown code blocks often chops off trailing parenthesis.

For example:

```
using Markdown: Markdown, MD

md = MD(Markdown.Code("julia", "gate_concurrence(U) = gate_concurrence(weyl_chamber_coordinates(U)...)"))
```

will print

```
gate_concurrence(U) = gate_concurrence(weyl_chamber_coordinates(U)...
```

Notice that there is a parenthesis missing at the end.

For more details, see the notebook at https://gist.github.com/goerz/bb2f1589d2aa523ffa8ccd4d0addcd86

It is not clear to me whether this is a bug in IJulia or a bug in Jupyter.

Also reported as https://github.com/JuliaLang/IJulia.jl/issues/1111

**Comment by goerz** [comment]
> For more details, see the notebook at https://gist.github.com/goerz/bb2f1589d2aa523ffa8ccd4d0addcd86

Huh, it seems like this issue does appear in the renderer used by Github, or [NBViewer](https://nbviewer.org/gist/goerz/bb2f1589d2aa523ffa8ccd4d0addcd86)! Thus, I suppose this has to be an issue in the *frontend*, which most likely puts the problem with JupyterLab.

Here is a screenshot from JupyterLab running on my laptop, showing the problem:

![CleanShot 2024-04-04 at 17 28 54@2x](https://github.com/JuliaLang/IJulia.jl/assets/112306/1d4f020e-459d-4971-81b9-16fd0fdb7fd5)

**Comment by goerz** [comment]
Adding a space in front of the parenthesis, i.e.,

```
md = MD(Markdown.Code("julia", "gate_concurrence(U) = gate_concurrence(weyl_chamber_coordinates(U)... )"))
```

also displays properly

**Comment by goerz** [comment]
Full version information:

```
jupyter --version
IPython          : 8.17.2
ipykernel        : 6.26.0
ipywidgets       : 8.1.1
jupyter_client   : 8.6.0
jupyter_core     : 5.5.0
jupyter_server   : 2.10.0
jupyterlab       : 4.0.8
nbclient         : 0.8.0
nbconvert        : 7.11.0
nbformat         : 5.9.2
notebook         : 7.0.6
qtconsole        : not installed
traitlets        : 5.13.0
```

Running on macOS. The issue is not browser-dependent (Tried Safari, Chrome, Arc)

Also tested on Linux. Same problem. Checked font settings. The problem occurs with all fonts I tried (JuliaMono, Menlo, Courier).

A colleague (using Linux) was not able to reproduce the issue, with versions 

```
jupyter --version
Selected Jupyter core packages...
IPython          : 7.31.1
ipykernel        : 6.19.2
ipywidgets       : 7.6.5
jupyter_client   : 7.4.8
jupyter_core     : 5.1.1
jupyter_server   : 1.23.4
jupyterlab       : 3.5.3
nbclient         : 0.5.13
nbconvert        : 6.4.4
nbformat         : 5.7.0
notebook         : 6.5.2
qtconsole        : 5.3.2
traitlets        : 5.7.1
```

So this *may* be a regression in `jupyterlab` 3 → 4.

**Comment by goerz** [comment]
https://mybinder.org/v2/gist/goerz/bb2f1589d2aa523ffa8ccd4d0addcd86/ffc203d463102fb53b626f64b2ac3875db1d75cb?urlpath=lab%2Ftree%2FIJuliaMarkdownBug.ipynb confirms that this is a regression bug in JupyterLab 4.

Without the `environment.yml` file, Binder uses JupyterLab 3 and the parenthesis appears correct. With the [`environment.yml`](https://gist.github.com/goerz/bb2f1589d2aa523ffa8ccd4d0addcd86/ffc203d463102fb53b626f64b2ac3875db1d75cb#file-environment-yml) file that forces `jupyterlab=4.0.8`, the parenthesis disappears.

Similarly, my colleague updated their JupyterLab from 3 to 4 and the parenthesis went from being there to not being there.

Note that reproducing this does not require executing the notebook. Thus, it does not depend on the IJulia kernel. Simply viewing the notebook in JupyterLab 3 or JupyterLab 4 reproduces the error.

**Comment by krassowski** [comment]
It does not seem to be a generic problem with Markdown rendering because I do not see it in Markdown cells in JupyterLab nor when rendering Markdown from IPython (JupyterLab 4.0.12 and 4.2.0b0):

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/daf0b6ab-a2bc-4f00-bfb3-b2e977e0c332)

Looking at the notebook you shared, your code actually renders code block within Markdown:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/b61eaa01-e3b9-454c-af84-2c073e2a08db)

And that code block is syntax-highlighted as Julia language. So after changing my Markdown reproducer to:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/afd9e178-b68a-4a53-997a-f8d3ea63ec3c)

I can reproduce this in Markdown cells in 4.0.12:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/6bfe66e1-651e-4180-9e58-675bbe876fbd)

And in 4.2.0b0 which includes updated CodeMirror packages.

**Comment by krassowski** [comment]
Is the line in question a valid Julia syntax?

If so, maybe we should try https://www.npmjs.com/package/@plutojl/lang-julia implementation instead of using the legacy mode.

**Comment by goerz** [comment]
> Is the line in question a valid Julia syntax?

Yes, it is. This is just splatting: `weyl_chamber_coordinates` returns a tuple `(c1, c2, c3)`, and `gate_concurrence` takes three positional parameters. The `...` just expands the tuple into three separate parameters. In Python, this would be `gate_concurrence(*weyl_chamber_coordinates(U))`

**Comment by krassowski** [comment]
This is no longer reproducible in JupyterLab 4.6.0:

<img width="824" height="219" alt="Image" src="https://github.com/user-attachments/assets/3c1a9cf8-519d-4c56-b480-94d4c21a8a48" />

It was fixed in JupyterLab 4.5.6, I think by https://github.com/jupyterlab/jupyterlab/pull/18572
